### ISSUE TITLE
Softmax using multiple threads inhibiting parallel execution of forward passes (?)

### ISSUE THREAD

**Issue by JakobHavtorn** [pytorch/pytorch#4628]
Using Linux Mint 18.3 and pytorch 0.3.0, some `nn.Module` and `nn.functional` methods appear to use multiple cores/threads and inhibit parallel execution. 

At the basic level, I am interested in running multiple forward passes in parallel. This works perfectly on Mac OS 10.13.2 but on Linux Mint 18.3, I face the problem described above.

For a network like the one below, using `softmax` while executing multiple forward passes in parallel with `torch.multiprocessing.Process` results in a complete backlog. If changed to `relu`, the code executes fine.

If testing the softmax/relu situation on a single thread, it appears that the softmax takes up multiple threads on Linux while on Mac OS it does not. Code used to test this is

```
import time
import gym
import numpy as np
import torch
import torch.multiprocessing as mp
import torch.nn.functional as F
from torch import nn
from torch.autograd import Variable


class FFN(nn.Module):
    def __init__(self, observation_space, action_space, acti='relu'):
        super(FFN, self).__init__()
        assert hasattr(observation_space, 'shape') and len(
            observation_space.shape) == 1
        assert hasattr(action_space, 'n')
        in_dim = observation_space.shape[0]
        out_dim = action_space.n
        self.lin1 = nn.Linear(in_dim, 32)
        self.lin2 = nn.Linear(32, 64)
        self.lin3 = nn.Linear(64, 64)
        self.lin4 = nn.Linear(64, 32)
        self.lin5 = nn.Linear(32, out_dim)
        self.acti = acti

    def forward(self, x):
        x = F.relu(self.lin1(x))
        x = F.relu(self.lin2(x))
        x = F.relu(self.lin3(x))
        x = F.relu(self.lin4(x))
        if self.acti == 'softmax':
            x = F.log_softmax(self.lin5(x), dim=1)
        elif self.acti == 'relu':
            x = F.relu(self.lin5(x))
        return x


def gym_rollout(max_episode_length, model, random_seed, env, is_antithetic):
    """
    Function to do rollouts of a policy defined by `model` in given environment
    """
    # Reset environment
    state = env.reset()
    state = Variable(torch.from_numpy(state).float(),
                     requires_grad=True).unsqueeze(0)
    retrn = 0
    nsteps = 0
    done = False
    # Rollout
    while not done and nsteps < max_episode_length:
        # Choose action
        actions = model(state)
        action = actions.max(1)[1].data
        # Step
        state, reward, done, _ = env.step(action[0])
        retrn += reward
        nsteps += 1
        # Cast state
        state = Variable(torch.from_numpy(state).float(),
                         requires_grad=True).unsqueeze(0)


for acti in ['relu', 'softmax']:
    ts = time.clock()
    n = 10000
    env = gym.make('CartPole-v0')
    model = FFN(env.observation_space, env.action_space, acti)

    for i in range(n):
        gym_rollout(1000, model, 'dummy_seed', env, 'dummy_neg')

    tf = time.clock()
    print(acti + " total: " + str(tf-ts))
    print(acti + " per call: " + str((tf - ts) / n))

```

On Mac OS, this outputs

```
relu total: 21.906198999999997
relu per call: 0.0021906199
softmax total: 21.903843000000002
softmax per call: 0.0021903843
```

On Linux (faster computer) this outputs
```
relu total: 8.02111
relu per call: 0,000802111
softmax total: 41,709772
softmax per call: 0,0041709772
```
It does however not take 41 seconds of real time for the `softmax` but more like the 8 seconds it takes the `relu`. That is, it uses multiple threads.

**Comment by apaszke** [comment]
Sorry, but is the problem that every process tries to use all the cores and this makes it slower for you? You can regulate the number of threads PyTorch uses by setting `torch.set_num_threads(X)` or running your script with `OMP_NUM_THREADS=X MKL_NUM_THREADS=X`. Does this solve your problem?

**Comment by JakobHavtorn** [comment]
Using `torch.set_num_threads(1)` solves the problem.

I figure the problem was a deadlock arising as a result of creating a huge number of threads:

I create a number of threads where each thread should execute a series of forward passes. However, when the thread encounters the softmax, it spawns a number of child threads. As every thread did this, a huge number of threads was being spawned in total, which caused a deadlock. 

It is funny though that this occurs only on Linux and not Mac OS.

**Comment by Tetratrio** [comment]
I have experienced the same problem. I am gathering trajectories for policy gradients with multiple processes. When I moved from a continous actionspace to a discrete action space (and changed the final activation in my policy net from sigmoid to softmax) the sampling got super slow. Using torch.set_num_threads(1) solves the problem.

Using python 3.5 and pytorch 0.3 installed via pip. Running on Ubuntu 16.04 on my home PC. No docker or similar things used.

**Comment by alexholdenmiller** [comment]
Can you add a comment to http://pytorch.org/docs/master/notes/multiprocessing.html about `torch.set_num_threads`? I was trying to do hogwild training, and if I had an nn.Module subclass which was loaded from file, without setting `torch.set_num_threads(1)` softmax was causing a deadlock every single time (even with only two threads).
For some reason this didn't happen when the model was initialized from scratch.

**Comment by apaszke** [comment]
@alexholdenmiller can you please send us a repro script? This shouldn't happen

**Comment by ezyang** [comment]
I'm... going to optimistically hope that we have fixed this
